Added CRD and design documents

shtripat opened the pull request:
monstorak-issue-id: ceph/ceph-mixins/issues/6
monstorak-issue-id: ceph/ceph-mixins/issues/22
Signed-off-by: Shubhendu <shtripat@redhat.com>

Commit:
Added CRD and design documents

monstorak-issue-id: ceph/ceph-mixins/issues/6
monstorak-issue-id: ceph/ceph-mixins/issues/22
Signed-off-by: Shubhendu <shtripat@redhat.com>

shtripat commented:
@umangachapagain as discussed, sending this PR. Please check.

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
 
-This operator takes care of deployment of mixins of storage system in Openshift and Kubernetes. The mixins are the mechanism for adding alerting rules, grafana dashboards and recording rules for prometheus in Openshift and Kubernetes.
+This operator takes care of deployment of mixins of storage system in
+kubernetes. The mixins are the mechanism for adding alerting rules,
```
I do not agree with this statement.
`Monitoring-Mixin` is not a mechanism, rather an abstraction/framework that makes it easier for developers to collaborate on common set of features (i.e prometheusRule, Grafana Dashboard Config etc.) and increases reusability etc.

shtripat replied:
Ok. Will try re-phrasing this sentence.

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
 # Monitoring operator for storages
```
Maybe, **Operator for Storage Monitoring Mixin** ?

shtripat replied:
Ack

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
+kubernetes. The mixins are the mechanism for adding alerting rules,
+grafana dashboards and recording rules for prometheus in kubernetes.
+
+# Deploy Operator with mixins
```
Like I said before, ~Deploy Operator with mixins~. **Deploy Mixin with Operator**

shtripat replied:
Ack

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
+grafana dashboards and recording rules for prometheus in kubernetes.
+
+# Deploy Operator with mixins
+To deploy the operator simply deploy to your cluster. The set of steps
```
These steps will change. There will only be one copy of deployment assets as same things work for both.

shtripat replied:
When is these steps finalized? Once done, will change this section.

umangachapagain replied:
It is finalized. 
Create a namespace where operator will be deployed. Then,
1. create crd
2. create serviceaccount
3. create role
4. create rolebinding
5. operator
6. storagealert cr

shtripat replied:
Ack, will update the PR accordingly

shtripat replied:
I feel you dont plan to have /deploy/kubernetes/*.yaml anymore.
Do we have a yaml for creating namespace?

umangachapagain replied:
We don't.

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
+# Deploy Operator with mixins
+To deploy the operator simply deploy to your cluster. The set of steps
+to be followed for deployment of monstorak operator and related artifacts,
+follow the below steps. These steps involve creation of namespace, cluster
```
We will be creating `role` and `roleBinding` not `cluster-*`

shtripat replied:
Ack

umangachapagain replied:
~cluster role~ -> `role`

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
+to be followed for deployment of monstorak operator and related artifacts,
+follow the below steps. These steps involve creation of namespace, cluster
+role, role binding and deployment of operator, its custom resources and
+finally the storage alerting rules.
```
You do not have to deploy any alerting rules. This is what the operator has to do.

shtripat replied:
ok, will remove this

shtripat replied:
I will change the statement as `The set of steps performed by operator involve
creation of namespace, cluster role, role binding, its custom resources and
finally the storage alerting rules.`

umangachapagain replied:
same as above.

shtripat replied:
> same as above.

You mean to remove deployment of alerting rules here?

umangachapagain replied:
https://github.com/monstorak/monstorak/pull/31#discussion_r285861164

shtripat replied:
Done

umangachapagain commented on README.md:
```diff
@@ -1,3 +1,23 @@
+oc create -f deploy/kubernetes/03_cluster-rolebinding.yaml
+oc create -f deploy/kubernetes/04_operator.yaml
+oc create -f deploy/kubernetes/05_storageAlerts.yaml
+oc create -f jsonnet/manifests/ceph-prometheus-rules.yaml
```
If we are doing this, we wouldn't need first 5 steps.

shtripat replied:
Ah ok, will remove

umangachapagain commented on README.md:
````diff
@@ -1,3 +1,23 @@
+oc create -f jsonnet/manifests/ceph-prometheus-rules.yaml
+```
+
+The above steps consider sample deployment of ceph alerting rules.
````
There is no sample. Operator deploys what we mark as release ready. 
Can remove this line.

shtripat replied:
Ok, will remove the comment

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+This document describes the set of [Custom Resource Definitions
+(CRDs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+that will be used to deploy mixins for a storage system. The actual implementation
```
~that will be used to deploy mixins for a storage system. The actual implementation
of the resources described here will be phased in during development. The
purpose of this document is to provide the overall structure, ensuring the end
result provides necessary configurability in a user-friendly manner.~

`that will be used to deploy mixin and ensure user-friendly configuration options  for monitoring storage systems.`

shtripat replied:
Ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+This document describes the set of [Custom Resource Definitions
```
~the set~ -> `a set`

shtripat replied:
Ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+
+# Overview
+
+A single Monstorak operator can take care of deployment of mixins for one or
```
~A single Monstorak operator can take care~ -> `An instance of Monstorak operator takes care`

shtripat replied:
Ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+# Overview
+
+A single Monstorak operator can take care of deployment of mixins for one or
+more storage systems. Each storage cluster should already be deployed within
```
~Each storage cluster should already be deployed within
the kubernetes cluster, or the storage cluster could be running on hosts outside
of kubernetes cluster. The capabilities of the operator remains same in both
the modes of the deployment of storage cluster and the same set of CRDs should
be used for both the cases.~ ->
It is expected that,
1. deployment of the storage cluster to be monitored is taken care by the user, and
2. prometheus can scrape metrics from this cluster.

shtripat replied:
Ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+the modes of the deployment of storage cluster and the same set of CRDs should
+be used for both the cases.
+
+A given deployment configuration for storage systems contains a Custom Resource
```
~A given deployment configuration for storage systems contains a Custom Resource
which contains configurations options like whether alerts are enabled, if
dashboards are enabled, name of the storage system and version details of storage
system.~

shtripat replied:
Ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+dashboards are enabled, name of the storage system and version details of storage
+system.
+
+![Hierarchy of monstorak custom resources](crd_hierarchy.dot.png)
```
There is no `Storage Monitoring Plan`.

shtripat replied:
Ok, will remove this. I felt still we are using StorageMonitoringPlan CR.

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+# Custom Resource Definitions
+
+This section describes the fields in each of the custom resources.
+All CRs live within the `mixins.monstorak.org` group and have version
```
~mixins.monstorak.org~ -> `alerts.monstorak.org`

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,69 @@
+
+### *Singular: StorageAlert, Plural:StorageAlerts*
+
+The StorageAlert custom resource defines a storage system for which mixins could
```
~The StorageAlert custom resource defines a storage system for which mixins could
be deployed using the operator. This CR captures the names of the storage system,
its supported version details, namespace in which alerts should be deployed.~

This CR defines configuration that is used by the operator to deploy monitoring mixin for the storage systems.

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
````diff
@@ -0,0 +1,69 @@
+its supported version details, namespace in which alerts should be deployed.
+
+```YAML
+apiVersion: mixins.monstorak.org/v1alpha1
````
~mixins.monstorak.org~ -> `alerts.monstorak.org`

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/CRD_design.md:
````diff
@@ -0,0 +1,69 @@
+
+```YAML
+apiVersion: mixins.monstorak.org/v1alpha1
+kind: StorageAlert
````
```
kind: StorageAlert
spec:
  - storage: "ceph"
    - version: "mimic"
      alert: false
    - version: "nautilus"
      alert: true
      prometheus:
      - namespace: "monitoring"
      - label: "prometheus:k8s"
```

I thought this was the proposed structure. @aruniiird Looks like even you are using older CRD structure.

shtripat replied:
@aruniiird if you confirm, I will change to this.
@umangachapagain I checked from `_types.go` and found this so added

umangachapagain commented on docs/Developers/Design/CRD_design.md:
````diff
@@ -0,0 +1,69 @@
+  - prometheusNamespace: "monitoring"
+```
+
+## **StorageMonitoringPlan**
````
Anything from here onward is not required.

shtripat replied:
ack, will remove this

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,42 @@
+This document describes the actions that have to be executed successfully for
+monstorak's CRDs to be considered reconciled.
+
+![ActionGraph](workflow.dot.png)
```
Diagram needs to be updated. Remove StorageMonitoringPlan.

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,42 @@
+are already available.
+
+## Storage cluster deployed
+The ceph mixins (alerting rules) are meant for storage, so the operator expects
```
Not specific to ceph. Make it generic.

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,42 @@
+![Rook](https://github.com/rook/rook/) takes care of deployment of ceph cluster.
+
+## Monitoring Enabled For Storage
+The namespace in which storage is deployed, monitoring needs to be enabled for
```
These steps are performed by the operator already and is OCP specific case which we should not promote.
Rather we should say that `Prometheus should be able to scrape metrics form Storage Cluster`

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
````diff
@@ -0,0 +1,42 @@
+- oc policy add-role-to-user view system:serviceaccount:openshift-monitoring:prometheus-k8s -n <storage-namespace>
+```
+
+# Deploy storage mixin
````
Looks like a part of pre-requisites. Needs a clear separator.

shtripat replied:
Earlier you only mentioned that this should be parallel to pre-requisites so made this way.

shtripat replied:
Also this is top level heading so why you feel, it looks like part of pre-requisites. While viewing the document it would look different.

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
````diff
@@ -0,0 +1,42 @@
+```
+
+# Deploy storage mixin
+Mixins are mechanism to write alerting rules and grafana dashboard templates in
````
~Mixins are mechanism~

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,42 @@
+
+# Deploy storage mixin
+Mixins are mechanism to write alerting rules and grafana dashboard templates in
+kubernetes. The project ![ceph-mixins](https://github.com/ceph/ceph-mixins/)
```
Should be generic. Can use `ceph` as an example or even `gluster`

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,42 @@
+templates, which ultimately needs to be compiled into prometheus rules YAML and
+grafana dashboard JSON templates.
+
+This operator take care of deploying the prometheus alerting rules generated out
```
Should be used as an example use case.

shtripat replied:
ack

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,42 @@
+This operator take care of deploying the prometheus alerting rules generated out
+of ![ceph-mixins](https://github.com/ceph/ceph-mixins/) in a kubernetes cluster.
+
+The monstorak operator consists of two controllers. The first one takes care of
```
~The monstorak operator consists of two controllers. The first one takes care of
the instanciating/reconciling custom resources like StorageCatalog,
StorageAlert and StorageDasboard. Once custom resources are available, the
second controller forms storage monitoring plans and executes them one by one
based on how many storage system and what artificats to be deployed.~

shtripat replied:
ack

Commit:
Incorporated review comments

monstorak-issue-id: ceph/ceph-mixins/issues/6
monstorak-issue-id: ceph/ceph-mixins/issues/22
Signed-off-by: Shubhendu <shtripat@redhat.com>

umangachapagain commented on docs/Developers/Design/CRD_design.md:
```diff
@@ -0,0 +1,39 @@
+
+An instance of monstorak operator takes care of deployment of mixins for one or
+more storage systems. It is expected that -
+1. deployment of the storae cluster to be monitored is taken care by the user
```
~storae~ -> `storage`

shtripat replied:
Done

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,32 @@
+of a ceph storage cluster.
+
+## Monitoring Enabled For Storage
+Prometheus should be able to scrape metrics form storage cluster.
```
~form~ -> `from`

shtripat replied:
Done

umangachapagain commented on docs/Developers/Design/deploy_procedure_design.md:
```diff
@@ -0,0 +1,32 @@
+grafana dashboard JSON templates.
+
+This operator take care of deploying the prometheus alerting rules generated out
+of mixins projects for storage systems like ![ceph-mixins](https://github.com/ceph/ceph-mixins/)
```
`storage systems like ceph`.

shtripat replied:
Ack

umangachapagain commented on README.md:
````diff
@@ -1,3 +1,24 @@
+below commands -
+
+```
+oc create -f deploy/serviceaccount.yaml
````
Create CRD is missing.

shtripat replied:
Done

cloudbehl commented:
PR looks good to me!

@umangachapagain do you see anything else which needs to done?
